### PR TITLE
Added network flow logging

### PR DESCRIPTION
**Pull request by kim-mskw**
# Pull Request

## Description
- added option for flow logging 
- implemented as new output pf function clear()

## Testing
Tested with multiple clearing algorithms with local and timescale as well as csv output

## Checklist
Please check all applicable items:

- [x] Code changes are sufficiently documented (docstrings, inline comments, `doc` folder updates)
- [x] New unit tests added for new features or bug fixes
- [x] Existing tests pass with the changes
- [x] Reinforcement learning examples are operational (for DRL-related changes)
- [x] Code tested with both local and Docker databases
- [x] Code follows project style guidelines and best practices
- [ ] Changes are backwards compatible, or deprecation notices added
- [x] New dependencies added to `pyproject.toml`
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included
- [x] Consent to release this PR's code under the GNU Affero General Public License v3.0

## Additional Notes (if applicable)
[Any additional information, concerns, or areas you want reviewers to focus on]

## Screenshots (if applicable)
[Add screenshots to demonstrate visual changes]

### CODE REVIEW

**Comment by kim-mskw** [comment]
I’d like to enable the option of logging the flows within the model network to the output roll. To achieve this, I’ve integrated a message from the market roll to the output roll by sending a designated message specifically for the flow data.

There’s also the option to log the flows in the market meta since we already have this data for the specific nodes. However, from a processing standpoint, I’d argue that a separate table for flow logging might be easier to manage and we omit one processing step of the flow data while simulating.

Before proceeding with this, I wanted to get your opinion on the approach?

**Review by maurerle — COMMENTED** [review]
In general, this is on the right track :)

File: assume/common/outputs.py [hunk 1]
```diff
@@ -569,3 +574,29 @@ def get_sum_reward(self):
         rewards_by_unit = np.array(rewards_by_unit)
 
         return rewards_by_unit
+
+    def write_flows(self, data: any):
```
Comment (maurerle) [inline assume/common/outputs.py:578]: data looks like `dict[tuple[datetime, str], float]`, is this correct?

**Review by maurerle — COMMENTED** [review]
Looks fine in general, though I would first like to merge the mango 2 PR.

In general it helps to have a second mechanism implement the flows as well, to see if one found a suitable abstraction.
But I am fine with merging this and adapting afterwards

File: assume/markets/clearing_algorithms/nodal_pricing.py [hunk 2]
```diff
@@ -175,8 +175,11 @@ def clear(
             meta.extend(
                 calculate_network_meta(network=nodal_network, product=product, i=i)
             )
+        # wirte network flows here if applicable
+        flows = []
```
[the remark this answers is not in the document]
Comment (kim-mskw) [reply assume/markets/clearing_algorithms/nodal_pricing.py:179]: Done :)

File: docs/source/release_notes.rst [hunk 3]
```diff
@@ -86,6 +86,7 @@ v0.4.0 - latest release (8th August 2024)
 - Examples have been restructed to easier orientation and understanding: example_01.. cover all feature demonstration examples,
   example_02.. cover all learning examples, example_03.. cover all full year examples
 - Added the option of integrating different actor network architectures to the reinforcement learning algorithm, currently a multilayer perceptron (mlp) and long short-term memory (lstm) are implemented
+- Added storing of network flows for complex clearing
```
[the remark this answers is not in the document]
Comment (kim-mskw) [reply docs/source/release_notes.rst:93]: Adjusted it for nodal clearing from pypsa as well

File: assume/common/outputs.py (hunk 1 above)
Comment (kim-mskw) [reply assume/common/outputs.py]: That is correct. This is the output of promo. I decided to leave it like that and do the data format handling in the output role to keep the market role as decluttered as possible.

**Comment by kim-mskw** [comment]
@maurerle, what do you think about handling format in the output role? Should I rather do that immediately in the clearing?

**Review by maurerle — APPROVED** [review]
> @maurerle, what do you think about handling format in the output role? Should I rather do that immediately in the clearing?

It is fine to do it in the output role, as long as the format is not weird but `dict[tuple[datetime, str], float]` is very fine.
Processing in the output role is preferred, as this can be run in an additional process (in the future on all plattforms).

The PR looks fine for me, I will update it to fix the tests.

**Pull request merged**